File: JavaAndScala/src/main/scala/org/lang/scala/conclusion/bignumber/TestBigNumber.scala
package org.lang.scala.conclusion.bignumber

/**
 * 	This is a stand-alone object to test BigDecimal and BigInt
 * 	-- BigDecimal can be used to ensure the precision of number calculation
 *  -- BigInt can be used to do mod and mod-inverse calculation
 *  -- BigDecimal and BigInt can represent the calculated result in any order of magnitude
 *  
 * 	@author VinceYuan
 */
object TestBigNumber {
  
  /*	Necessary instance variables	*/
  private val d1 = 2.0d; private val d2 = 1.1d; private val d3 = 3.0d
  private val f1 = 0.08f; private val f2 = 0.01f; private val f3 = 0.2f
  
  /**
   * 	This is a main method for execution
   */
  def main(args: Array[String]): Unit = {
    testNumberCalculation()
    println()
    testAccurateCalculation()
    println()
    testModAndModInverseCalculation()
    println()
    testBigNumberRepresentation()
    println()
    testOtherMethods()
  }
  
	/**
	 * 	There are problems regarding number calculation
	 * 	-- Double-typed and float-typed calculation maybe inaccurate
	 */
  private def testNumberCalculation(): Unit = {
    
		/*	Calculation of Double type maybe inaccurate	 */
		println("Calculation of Double type maybe inaccurate. For examples:")
		println("Substraction: " + d1 + " - " + d2 + " = " + (d1 - d2))
		println("Multiplication: " + d3 + " * " + d2 + " = " + (d3 * d2))
		
		/*	Calculation of Float type maybe inaccurate	*/
		println("\nCalculation of Float type maybe inaccurate. For exmaples:")
		println("Addition: " + f1 + " + " + f2 + " = " + (f1 + f2))
		println("Division: " + f1 + " / " + f3 + " = " + (f1 / f3))
  }

	/**
	 * 	This is a method to test accurate calculation
	 * 
	 * 	BigDecimal can be used to:
	 * 	-- Address the calculation inaccuracy
	 */
  private def testAccurateCalculation(): Unit = {

		/*	
		 * 	Use BigDecimal to do precise calculation
		 *  -- Make sure the argument is (converted to) String type	
		 */
		val d1_bd = BigDecimal(d1)
		val d2_bd = BigDecimal(String.valueOf(d2))      // This way is recommended due to declaration consistency of different data types
		val d3_bd = BigDecimal.valueOf(d3)
		val f1_bd = BigDecimal.decimal(f1)
		val f2_bd = BigDecimal(String.valueOf(f2))
		val f3_bd = BigDecimal(String.valueOf(f3))
		System.out.println("Calculation of BigDecimal type is accurate:")
		System.out.println("Subtraction: " + d1_bd + " - " + d2_bd + " = " + (d1_bd - d2_bd).doubleValue())
		System.out.println("Multiplication: " + d3_bd + " * " + d2_bd + " = " + (d3_bd * d2_bd).doubleValue())
		System.out.println("Addition: " + f1_bd + " + " + f2_bd + " = " + (f1_bd + f2_bd).floatValue())
		System.out.println("Division: " + f1_bd + " / " + f3_bd + " = " + (f1_bd / f3_bd).floatValue())
  }
  
  /**
   * 	This is a method to test mod and mod-inverse calculation
   * 
   *  BigInt can be used to:
	 * 	-- Do mod and mod-inverse calculation
   */
  private def testModAndModInverseCalculation(): Unit = {
    
    /*	Initialize two integers	 */
    val i = 5
    val j = 12
    
    /*	Do mod and mod-inverse calculation	*/
    val mod_1 = BigInt(i).mod(BigInt(j))
    val mod_2 = BigInt(j).mod(BigInt(i))
    val modInverse_1 = BigInt(i).modInverse(BigInt(j))
    val modInverse_2 = BigInt(j).modInverse(BigInt(i))
    println("Mod and Mod-Inverse calculation through BigInt:")
    println(s"$i mod $j: $mod_1")
    println(s"$j mod $i: $mod_2")
    println(s"$i mod-inverse $j: $modInverse_1")
    println(s"$j mod-inverse $i: $modInverse_2")
  }
  
  /**
	 * 	This is a method to test big number representation
	 * 
	 * 	BigDecimal and BigInt can be used to:
	 *  -- Represent an integer with any order of magnitude
	 *  -- Notice that when a number becomes very large, BigDecimal will adopt scientific notation while BigInt will not
	 *     -- This will make some undesirable result. E.g., compare 2 computational results between BigDecimal and BigInt
	 */
  private def testBigNumberRepresentation(): Unit = {
    
    /*	Use BigDecimal to do (1.5 * 2.5 * ... * 999.5 * 1000.5)	 */
		var sum_bd = BigDecimal(String.valueOf(1))
		for (i <- 1.5 to 1000.5 by 1) {
			sum_bd = sum_bd * BigDecimal(String.valueOf(i))
		}
		println("The (1.5 * 2.5 * ... * 999.5 * 1000.5) result using BigDecimal:\n" + sum_bd)
		println("The digits of the result: " + sum_bd.precision)
		println("The digits of decimal part: " + sum_bd.scale)
		
		/*	Use BigDecimal and BigInteger to do 1000 factorial and compare the results	*/
		sum_bd = BigDecimal(String.valueOf(1))
		for (i <- 1 to 1000) {
			sum_bd = sum_bd * BigDecimal(String.valueOf(i))
		}
		println("\nThe 1000 factorial result using BigDecimal:\n" + sum_bd)
		println("The digits of the result: " + sum_bd.precision)
		println("The digits of decimal part: " + sum_bd.scale)
		
		/*	Use BigInt to do 1000 factorial	 */
		var sum_bi = BigInt(String.valueOf(1))
		for (i <- 1 to 1000) {
			sum_bi = sum_bi * BigInt(String.valueOf(i))
		}
		println("\nThe 1000 factorial result using BigInt:\n" + sum_bi)
		println("The digits of the result: " + sum_bd.precision)
		println("The digits of decimal part: " + sum_bd.scale)
		
		/*	Compare the calculated results from BigDecimal and BigInteger	*/
		if (sum_bd != null && sum_bi != null) {
			println("\nIf these two 1000 factorial results are equal: " + sum_bi.equals(sum_bd.toBigInt()) + " (due to the scientific notation of BigDecimal in Scala)")
		}
  }
  
  /**
   * 	This is a method to test other methods of BigDecimal and BigInt
   */
  private def testOtherMethods(): Unit = {
 
		/*	Here is to test BigDecimal methods	 */
		val bd = BigDecimal(String.valueOf(921.2))
		println("Below is big decimal:")
		println("The scale of \"" + bd + "\": " + bd.scale)
		println("The precision of \""+ bd + "\": " + bd.precision)
//		println("The unscaled value of \"" + bd + "\": " + bd.unscaledValue)
//		println("The negate of \"" + bd + "\": " + bd.negate)
//		println("The plus of \"" + bd + "\": " + bd.plus)
		println("The sigum of \"" + bd + "\": " + bd.signum)
		println("The ulp of \"" + bd + "\": " + bd.ulp)
		
		/*	Here is to test BigInt methods	*/
		val bi = BigInt(String.valueOf(234))
		val bi2 = BigInt(String.valueOf(432))
		System.out.println("\nBelow is big decimal:")
		System.out.println("The bit count of \"" + bi + "\": " + bi.bitCount)
		System.out.println("The NOT result of \"" + bi + "\": " + (- (bi + 1)))
		System.out.println("The AND result of \"" + bi + "\" and \"" + bi2 + "\": " + (bi & bi2))
		System.out.println("The OR result of \"" + bi + "\" and \"" + bi2 + "\": " + (bi | bi2))
		System.out.println("The XOR result of \"" + bi + "\" and \"" + bi2 + "\": " + ((bi | bi2) &~ (bi & bi2)))
  }
}

File: JavaAndScala/src/main/java/com/java/se/conclusion/bignumber/TestBigNumber.java
package com.java.se.conclusion.bignumber;

import java.math.BigDecimal;
import java.math.BigInteger;

/**
 * 	This is a class to test BigDecimal and BigInteger
 * 	-- BigInteger and BigDecimal can be used to ensure the precision of number calculation
 *  -- BigInteger and BigDecimal can represent the calculated result in any order of magnitude
 *  
 * @author VinceYuan
 *
 */
public class TestBigNumber {

	/*	Necessary instance variables	*/
	private double d1, d2, d3;
	private float f1, f2, f3;
	private BigDecimal sum_bd;
	private BigInteger sum_bi;
	
	/**
	 * 	This is a main method for execution
	 * 
	 * @param args
	 */
	public static void main(String[] args) {	
		TestBigNumber testBigNumber = new TestBigNumber();
		testBigNumber.testPreparation();
		System.out.println("Here tests number calculation:");
		testBigNumber.testNumberCalculation();
		System.out.println("\nHere tests accurate calculation:");
		testBigNumber.testAccurateCalculation();
		System.out.println("\nHere tests big number representation:");
		testBigNumber.testBigNumberRepresentation();
		System.out.println("\nHere tests other methods:");
		testBigNumber.testOtherMethods();
	}
	
	/**
	 * 	This is a method for test preparation
	 */
	private void testPreparation() {
		d1 = 2.0; d2 = 1.1; d3 = 3.0;
		f1 = 0.08f; f2 = 0.01f; f3 = 0.2f;
	}
	
	/**
	 * 	There are problems regarding number calculation
	 * 	-- Double-typed and float-typed calculation maybe inaccurate
	 */
	private void testNumberCalculation () {
		
		/*	Calculation of Double type maybe inaccurate	 */
		System.out.println("Calculation of Double type maybe inaccurate. For examples:");
		System.out.println("Substraction: " + d1 + " - " + d2 + " = " + (d1 - d2));
		System.out.println("Multiplication: " + d3 + " * " + d2 + " = " + (d3 * d2));
		
		/*	Calculation of Float type maybe inaccurate	*/
		System.out.println("Calculation of Float type maybe inaccurate. For exmaples:");
		System.out.println("Addition: " + f1 + " + " + f2 + " = " + (f1 + f2));
		System.out.println("Division: " + f1 + " / " + f3 + " = " + (f1 / f3));
	}
	
	/**
	 * 	BigDecimal can be used to:
	 * 	-- Address the calculation inaccuracy
	 */
	private void testAccurateCalculation () {
		
		/*	
		 * 	Use BigDecimal to do precise calculation
		 *  -- Make sure the argument is (converted to) String type	
		 */
		BigDecimal d1_bd = new BigDecimal(Double.toString(d1));
		BigDecimal d2_bd = new BigDecimal(String.valueOf(d2));		// This way is recommended due to declaration consistency of different data types				
		BigDecimal d3_bd = BigDecimal.valueOf(d3);
		BigDecimal f1_bd = new BigDecimal(Float.toString(f1));
		BigDecimal f2_bd = new BigDecimal(String.valueOf(f2));
		BigDecimal f3_bd = new BigDecimal(String.valueOf(f3));
		System.out.println("Calculation of BigDecimal type is accurate:");
		System.out.println("Subtraction: " + d1_bd + " - " + d2_bd + " = " + d1_bd.subtract(d2_bd).doubleValue());
		System.out.println("Multiplication: " + d3_bd + " * " + d2_bd + " = " + d3_bd.multiply(d2_bd).doubleValue());
		System.out.println("Addition: " + f1_bd + " + " + f2_bd + " = " + f1_bd.add(f2_bd).floatValue());
		System.out.println("Division: " + f1_bd + " / " + f3_bd + " = " + f1_bd.divide(f3_bd).floatValue());
	}
	
	/**
	 * 	BigDecimal and BigInteger can be used to:
	 *  -- Represent an integer with any order of magnitude
	 */
	private void testBigNumberRepresentation() {
		
		/*	Use BigDecimal to do (1.5 * 2.5 * ... * 999.5 * 1000.5)	 */
		sum_bd = new BigDecimal(String.valueOf(1));
		for (double i = 1.5; i <= 1000.5; i++) {
			sum_bd = sum_bd.multiply(new BigDecimal(String.valueOf(i)));
		}
		System.out.println("The (1.5 * 2.5 * ... * 999.5 * 1000.5) result using BigDecimal:\n" + sum_bd);
		System.out.println("The digits of the result: " + sum_bd.precision());
		System.out.println("The digits of decimal part: " + sum_bd.scale());
		
		/*	Use BigDecimal and BigInteger to do 1000 factorial and compare the results	*/
		sum_bd = new BigDecimal(String.valueOf(1));
		for (int i = 1; i <= 1000; i++) {
			sum_bd = sum_bd.multiply(BigDecimal.valueOf(i));
		}
		System.out.println("\nThe 1000 factorial result using BigDecimal:\n" + sum_bd);
		System.out.println("The digits of the result: " + sum_bd.precision());
		System.out.println("The digits of decimal part: " + sum_bd.scale());
		
		/*	Use BigInteger to do 1000 factorial	 */
		sum_bi = new BigInteger(String.valueOf(1));
		for (int i = 1; i <= 1000; i++) {
			sum_bi = sum_bi.multiply(BigInteger.valueOf(i));
		}
		System.out.println("\nThe 1000 factorial result using BigInteger:\n" + sum_bi);
		System.out.println("The digits of the result: " + sum_bd.precision());
		System.out.println("The digits of decimal part: " + sum_bd.scale());
		
		/*	Compare the calculated results from BigDecimal and BigInteger	*/
		if (sum_bd != null && sum_bi != null) {
			System.out.println("\nIf these two 1000 factorial results are equal: " + (sum_bi.compareTo(sum_bd.toBigInteger()) == 0));		
		}
	}
	
	/**
	 * 	Test other methods of BigDecimal and BigInteger
	 */
	private void testOtherMethods() {
		
		/*	Here is to test BigDecimal methods	 */
		BigDecimal bd = new BigDecimal(String.valueOf(921.2));
		System.out.println("Below is big decimal:");
		System.out.println("The scale of \"" + bd + "\": " + bd.scale());
		System.out.println("The precision of \""+ bd + "\": " + bd.precision());
		System.out.println("The unscaled value of \"" + bd + "\": " + bd.unscaledValue());
		System.out.println("The negate of \"" + bd + "\": " + bd.negate());
		System.out.println("The plus of \"" + bd + "\": " + bd.plus());
		System.out.println("The sigum of \"" + bd + "\": " + bd.signum());
		System.out.println("The ulp of \"" + bd + "\": " + bd.ulp());
		
		/*	Here is to test BigInteger methods	*/
		BigInteger bi = new BigInteger(String.valueOf(234));
		BigInteger bi2 = new BigInteger(String.valueOf(432));
		System.out.println("\nBelow is big decimal:");
		System.out.println("The bit count of \"" + bi + "\": " + bi.bitCount());
		System.out.println("The NOT result of \"" + bi + "\": " + bi.not());
		System.out.println("The AND result of \"" + bi + "\" and \"" + bi2 + "\": " + bi.and(bi2));
		System.out.println("The OR result of \"" + bi + "\" and \"" + bi2 + "\": " + bi.or(bi2));
		System.out.println("The XOR result of \"" + bi + "\" and \"" + bi2 + "\": " + bi.xor(bi2));
	}
}
